File: backend/personnel_management_v2/src/main/java/com/project2/personnel_management_v2/Services/AccountService.java
package com.project2.personnel_management_v2.Services;


import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.HttpServerErrorException;
import org.springframework.web.client.RestTemplate;

import java.util.Collections;

@Service
public class AccountService
{
    private final RestTemplate template = new RestTemplate();

    @Value("${sp.url}")
    private String url;

    @Value("${sp.username}")
    private String username;

    @Value("${sp.password}")
    private String pass;

    public ResponseEntity<String> createAccount(String account)
    {

       try
       {
           HttpHeaders headers = new HttpHeaders();
           headers.setContentType(MediaType.APPLICATION_JSON);
           headers.setBasicAuth(username,pass);
           headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON));
           HttpEntity<String> entity = new HttpEntity<>(account, headers);

           ResponseEntity<String> response = template
                   .exchange(url + "Accounts",
                           HttpMethod.POST,
                           entity,
                           String.class);

           return response;

       }catch (HttpServerErrorException e)
       {
           if(!e.getResponseBodyAsString().contains("detail"))
           {
                return new ResponseEntity<String>(HttpStatus.CREATED);
           }
           throw e;
       }
    }

    public String getAllAccounts()
    {
        HttpHeaders headers = new HttpHeaders();
        headers.setBasicAuth(username, pass);
        HttpEntity<String> entity = new HttpEntity<>(headers);
        ResponseEntity<String> response = template.
                exchange(url + "Accounts",
                        HttpMethod.GET,
                        entity,
                        String.class);

        return response.getBody();
    }

    public String getAccountById(String accountId)
    {
        HttpHeaders headers = new HttpHeaders();
        headers.setBasicAuth(username, pass);
        HttpEntity<String> entity = new HttpEntity<>(headers);
        ResponseEntity<String> response = template.
                exchange(url + "Accounts/"+ accountId,
                        HttpMethod.GET,
                        entity,
                        String.class);

        return response.getBody();
    }

    public String deleteAccount(String accountId)
    {
        HttpHeaders headers = new HttpHeaders();
        headers.setBasicAuth(username, pass);
        HttpEntity<String> entity = new HttpEntity<>(headers);
        ResponseEntity<String> response = template.
                exchange(url + "Accounts/" + accountId,
                        HttpMethod.DELETE,
                        entity,
                        String.class);

        return response.getBody();
    }

    public String updateAccount(String accountId, String body)
    {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.setBasicAuth(username, pass);
        headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON));
        HttpEntity<String> entity = new HttpEntity<>(body, headers);
        System.out.println(body);
        ResponseEntity<String> response = template.
                exchange(url + "Accounts/" + accountId,
                        HttpMethod.PUT,
                        entity,
                        String.class);

        return response.getBody();

    }
}

File: frontend/project-2-Angular/src/app/Services/account.service.ts
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { AuthService } from './auth.service';
import { AccountDetail } from '../Models/accountDetails'; // Import the 'Account' class

@Injectable()
export class AccountService {

  constructor(private http: HttpClient, private authService: AuthService) { }

  url: String = 'http://localhost:3000/';

  private getHeaders() {
    return this.authService.getHeaders();
  }

  getAllAccounts() {
    return this.http.get(this.url + 'api/accounts', { headers: this.getHeaders(), observe: 'response' });
  }

  getAccountById(id: string) {
    return this.http.get(this.url + 'api/accounts/' + id, { headers: this.getHeaders(), observe: 'response' });
  }

  createAccount(account: AccountDetail) {
    return this.http.post(this.url + 'api/accounts', account, { headers: this.getHeaders(), observe: 'response' });
  }

  deleteAccount(id: string) {
    return this.http.delete(this.url + 'api/users/' + id, { headers: this.getHeaders(), observe: 'response' });
  }

}
